Enable FTL on Windows

iangrunert opened the pull request:
#### dc849c8a2bb93c274ed63817818bda68ba05bb7e
<pre>
Enable FTL on Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=145366">https://bugs.webkit.org/show_bug.cgi?id=145366</a>

Reviewed by Yusuke Suzuki.

Disabling BBQ and OMG JIT for now, there&apos;s some edge cases which are
currently broken.

Fixes to testmasm, broken by previous sysv abi work.

B3 float mod tests were broken on Windows due to differences between
fmod(double, double) used within MathCommon fmodDouble, and the
fmod(float, float) used within the tests. Using fmodl in the tests so
they match behaviour with compiled B3.

The countLeadingZero implementation for testClz* tests had an off-by-one
error on Windows - using builtin __lzcnt64 and __lzcnt instead.

Had to precommit stack memory for testCallFunctionWithHellaArguments3 to
avoid accessing stack memory past the stack guard page.

* Source/JavaScriptCore/assembler/ProbeContext.cpp:
(JSC::Probe::flushDirtyStackPages):
* Source/JavaScriptCore/assembler/ProbeContext.h:
* Source/JavaScriptCore/assembler/testmasm.cpp:
(JSC::invoke):
(JSC::testStoreBaseIndex):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/air/opcode_generator.rb:
* Source/JavaScriptCore/b3/air/testair.cpp:
* Source/JavaScriptCore/b3/testb3.h:
(invoke):
* Source/JavaScriptCore/b3/testb3_2.cpp:
(testModArgFloat):
(testModArgsFloat):
(testModArgImmFloat):
(testModImmArgFloat):
(testModImmsFloat):
* Source/JavaScriptCore/b3/testb3_3.cpp:
(countLeadingZero):
* Source/JavaScriptCore/b3/testb3_4.cpp:
(testLoadFromFramePointer):
* Source/JavaScriptCore/b3/testb3_5.cpp:
(preCommitStackMemory):
(testCallFunctionWithHellaArguments3):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::ftlCalleeSaveRegisters):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::RegisterBinding::none):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSIMDV_V):
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::buildEntryBufferForCatchSIMD):
(JSC::Wasm::buildEntryBufferForCatchNoSIMD):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWin.cmake:

Canonical link: <a href="https://commits.webkit.org/280777@main">https://commits.webkit.org/280777@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f59e73a8fdb7ea5eafe12af537877c2ed337bff

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57593 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10068 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61215 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8038 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8226 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46634 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5704 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59623 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34640 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49760 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27500 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31417 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7054 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7041 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50685 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53371 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7327 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62895 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56835 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1507 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7420 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53894 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1513 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49775 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54002 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1282 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78596 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8590 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32750 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13036 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33836 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34920 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33581 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/testb3_3.cpp:
```diff
@@ -1690,7 +1690,11 @@ void testClzArg64(int64_t a)
+    CHECK_EQ(compileAndRun<unsigned>(proc, a), __lzcnt64(a));
+#else
     CHECK(compileAndRun<unsigned>(proc, a) == countLeadingZero(a));
+#endif
```
We should make `countLeadingZero` work on Windows too.

Constellation replied:
And we should explore why this is not correctly working on Windows.

Constellation replied:
> The countLeadingZero implementation for testClz* tests had an off-by-one
> error on Windows - using builtin __lzcnt64 and __lzcnt instead.

Using `__lzcnt64` etc. is not right. We should know why this does not work on Windows, and we need to fix it.

iangrunert replied:
Fixed - inside countLeadingZero we were left shifting a long (1L) instead of an unsigned long long (1ull) by the bitcount.

On debug builds this was using a 32 bit signed long, on release builds a 64 bit signed long. That gave us the off-by-one on release builds.

Constellation commented on Source/JavaScriptCore/b3/testb3_3.cpp:
```diff
@@ -1701,7 +1705,11 @@ void testClzMem64(int64_t a)
+    CHECK_EQ(compileAndRun<unsigned>(proc, &a), __lzcnt64(a));
+#else
     CHECK(compileAndRun<unsigned>(proc, &a) == countLeadingZero(a));
+#endif
```
Ditto

Constellation commented on Source/JavaScriptCore/b3/testb3_3.cpp:
```diff
@@ -1712,7 +1720,11 @@ void testClzArg32(int32_t a)
+    CHECK_EQ(compileAndRun<unsigned>(proc, a), __lzcnt(a));
+#else
     CHECK(compileAndRun<unsigned>(proc, a) == countLeadingZero(a));
+#endif
```
Ditto

Constellation commented on Source/JavaScriptCore/b3/testb3_3.cpp:
```diff
@@ -1723,7 +1735,11 @@ void testClzMem32(int32_t a)
+    CHECK_EQ(compileAndRun<unsigned>(proc, &a), __lzcnt(a));
+#else
     CHECK(compileAndRun<unsigned>(proc, &a) == countLeadingZero(a));
+#endif
```
Ditto

Constellation reviewed (approved):
r=me

iangrunert committed:
Enable FTL on Windows
https://bugs.webkit.org/show_bug.cgi?id=145366

Reviewed by Yusuke Suzuki.

Disabling BBQ and OMG JIT for now, there's some edge cases which are
currently broken.

Fixes to testmasm, broken by previous sysv abi work.

B3 float mod tests were broken on Windows due to differences between
fmod(double, double) used within MathCommon fmodDouble, and the
fmod(float, float) used within the tests. Using fmodl in the tests so
they match behaviour with compiled B3.

The countLeadingZero implementation for testClz* tests had an off-by-one
error on Windows - using builtin __lzcnt64 and __lzcnt instead.

Had to precommit stack memory for testCallFunctionWithHellaArguments3 to
avoid accessing stack memory past the stack guard page.

* Source/JavaScriptCore/assembler/ProbeContext.cpp:
(JSC::Probe::flushDirtyStackPages):
* Source/JavaScriptCore/assembler/ProbeContext.h:
* Source/JavaScriptCore/assembler/testmasm.cpp:
(JSC::invoke):
(JSC::testStoreBaseIndex):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/air/opcode_generator.rb:
* Source/JavaScriptCore/b3/air/testair.cpp:
* Source/JavaScriptCore/b3/testb3.h:
(invoke):
* Source/JavaScriptCore/b3/testb3_2.cpp:
(testModArgFloat):
(testModArgsFloat):
(testModArgImmFloat):
(testModImmArgFloat):
(testModImmsFloat):
* Source/JavaScriptCore/b3/testb3_3.cpp:
(countLeadingZero):
* Source/JavaScriptCore/b3/testb3_4.cpp:
(testLoadFromFramePointer):
* Source/JavaScriptCore/b3/testb3_5.cpp:
(preCommitStackMemory):
(testCallFunctionWithHellaArguments3):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::ftlCalleeSaveRegisters):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::RegisterBinding::none):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSIMDV_V):
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::buildEntryBufferForCatchSIMD):
(JSC::Wasm::buildEntryBufferForCatchNoSIMD):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWin.cmake:

Canonical link: https://commits.webkit.org/280777@main
